File: app/src/main/java/cn/devcxl/photosync/data/AppDatabase.java
package cn.devcxl.photosync.data;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import cn.devcxl.photosync.data.entity.PhotoEntity;
import cn.devcxl.photosync.data.entity.SyncDevice;

/**
 * @author devcxl
 */
@Database(entities = {SyncDevice.class, PhotoEntity.class}, version = 1, exportSchema = false)
public abstract class AppDatabase extends RoomDatabase {
    private static volatile AppDatabase INSTANCE;

    public abstract SyncDeviceDao syncDeviceDao();
    public abstract PhotoDao photoDao();

    public static AppDatabase getInstance(Context context) {
        if (INSTANCE == null) {
            synchronized (AppDatabase.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(), AppDatabase.class, "photosync.db")
                            .fallbackToDestructiveMigration()
                            .build();
                }
            }
        }
        return INSTANCE;
    }
}

File: app/src/main/java/cn/devcxl/photosync/data/AppDatabase.kt
package cn.devcxl.photosync.data

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import cn.devcxl.photosync.data.entity.PhotoEntity
import cn.devcxl.photosync.data.entity.SyncDevice

/**
 * @author devcxl
 */
@Database(entities = [SyncDevice::class, PhotoEntity::class], version = 2, exportSchema = false)
abstract class AppDatabase : RoomDatabase() {
    abstract fun syncDeviceDao(): SyncDeviceDao

    abstract fun photoDao(): PhotoDao


    companion object {
        @Volatile
        private var INSTANCE: AppDatabase? = null

        @JvmStatic
        fun getInstance(context: Context): AppDatabase {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: Room.databaseBuilder(
                    context.applicationContext,
                    AppDatabase::class.java,
                    "photosync.db"
                )
                    .fallbackToDestructiveMigration()
                    .build()
                    .also { INSTANCE = it }
            }
        }
    }
}
